### PR TITLE
Improve dataclasses documentation

### PR DESCRIPTION
**Pull request by Viicos**
## Change Summary

Closes https://github.com/pydantic/pydantic/issues/10679 along the way

## Checklist

* [ ] The pull request title is a good summary of the changes - it will be used in the changelog
* [ ] Unit tests for the changes exist
* [ ] Tests pass on CI
* [ ] Documentation reflects the changes where applicable
* [ ] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

File: docs/concepts/dataclasses.md [hunk 1]
```diff
@@ -25,26 +25,26 @@ User(id=42, name='John Doe', signup_ts=datetime.datetime(2032, 6, 21, 12, 0))
...
+*  [validators](#validators-and-initialization-hooks).
+*  The behavior with the [`extra`][pydantic.ConfigDict.extra] configuration value.
 
 You can use all the standard Pydantic field types. Note, however, that arguments passed to constructor will be copied in
 order to perform validation and, where necessary coercion.
 
-To perform validation or generate a JSON schema on a Pydantic dataclass, you should now wrap the dataclass
-with a [`TypeAdapter`][pydantic.type_adapter.TypeAdapter] and make use of its methods.
+To make use of the [various methods](./models.md#model-methods-and-properties) to validate and dump models, you can
+wrap the dataclass with a [`TypeAdapter`][pydantic.type_adapter.TypeAdapter] and make use of its methods.
```
Comment (Viicos) [inline docs/concepts/dataclasses.md:45]: The last _JSON dumping_ section was removed as this already explains it

File: docs/concepts/dataclasses.md [hunk 2]
```diff
@@ -100,52 +100,45 @@ print(TypeAdapter(User).json_schema())
...
 !!! note
-    Pydantic dataclasses support [`extra`][pydantic.config.ConfigDict.extra] configuration to `ignore`, `forbid`, or
-    `allow` extra fields passed to the initializer. However, some default behavior of stdlib dataclasses may prevail.
-    For example, any extra fields present on a Pydantic dataclass using `extra='allow'` are omitted when the dataclass
-    is `print`ed.
+    While Pydantic dataclasses support the [`extra`][pydantic.config.ConfigDict.extra] configuration value, some default
+    behavior of stdlib dataclasses may prevail. For example, any extra fields present on a Pydantic dataclass with
+    [`extra`][pydantic.config.ConfigDict.extra] set to `'allow'` are omitted in the dataclass' string representation.
 
 ## Nested dataclasses
```
Comment (Viicos) [inline docs/concepts/dataclasses.md:139]: Can we simplify this section and the following one with a single _Similarly to Pydantic models, nested dataclasses and generics are supported_ (and we refer to the relevant model documentation)? Imo the added value is quite low here, we would expect the examples to work anyway

File: docs/concepts/dataclasses.md [hunk 3]
```diff
@@ -25,26 +25,26 @@ User(id=42, name='John Doe', signup_ts=datetime.datetime(2032, 6, 21, 12, 0))
...
-Some differences between Pydantic dataclasses and `BaseModel` include:
+Some differences between Pydantic dataclasses and [`BaseModel`][pydantic.BaseModel] include:
 
-*  How [initialization hooks](#initialization-hooks) work
-*  [JSON dumping](#json-dumping)
+*  [validators](#validators-and-initialization-hooks).
+*  The behavior with the [`extra`][pydantic.ConfigDict.extra] configuration value.
 
 You can use all the standard Pydantic field types. Note, however, that arguments passed to constructor will be copied in
 order to perform validation and, where necessary coercion.
```
Comment (Viicos) [inline docs/concepts/dataclasses.md:42]: This is confusing, which _constructor_ are we talking about?

File: docs/concepts/dataclasses.md [hunk 4]
```diff
@@ -239,15 +233,14 @@ except pydantic.ValidationError as e:
...
+### Usage of stdlib dataclasses with `BaseModel`
 
-Bear in mind that stdlib dataclasses (nested or not) are **automatically converted** into Pydantic
-dataclasses when mixed with `BaseModel`! Furthermore the generated Pydantic dataclass will have
-the **exact same configuration** (`order`, `frozen`, ...) as the original one.
+When a standard library dataclass is used within a Pydantic model, a Pydantic dataclass or a [`TypeAdapter`][pydantic.TypeAdapter],
+validation will be applied (and the [configuration](#dataclass-config) stays the same). This means that using a stdlib or a Pydantic
+dataclass as a field annotation is functionally equivalent.
 
 ```py
```
Comment (Viicos) [inline docs/concepts/dataclasses.md:242]: Example simplified a bit

File: docs/concepts/dataclasses.md [hunk 5]
```diff
@@ -390,123 +376,78 @@ print(pydantic.dataclasses.is_pydantic_dataclass(PydanticDataclass))
 #> True
 ```
 
-## Initialization hooks
+## Validators and initialization hooks
 
-When you initialize a dataclass, it is possible to execute code *before* or *after* validation
-with the help of the [`@model_validator`](validators.md#model-validators) decorator `mode` parameter.
+Validators also work with Pydantic dataclasses:
```
Comment (Viicos) [inline docs/concepts/dataclasses.md:381]: Example below moved from the `validators.md` file

File: docs/concepts/dataclasses.md [hunk 6]
```diff
@@ -25,26 +26,26 @@ User(id=42, name='John Doe', signup_ts=datetime.datetime(2032, 6, 21, 12, 0))
...
     For more information and discussion see
     [pydantic/pydantic#710](https://github.com/pydantic/pydantic/issues/710).
 
-Some differences between Pydantic dataclasses and `BaseModel` include:
+Some differences between Pydantic dataclasses and [`BaseModel`][pydantic.BaseModel] include:
 
-*  How [initialization hooks](#initialization-hooks) work
-*  [JSON dumping](#json-dumping)
+*  [validators](#validators-and-initialization-hooks).
+*  The behavior with the [`extra`][pydantic.ConfigDict.extra] configuration value.
```
Comment (sydney-runkle) [inline docs/concepts/dataclasses.md:40]:
```suggestion
*  [validators](#validators-and-initialization-hooks)
*  The behavior with the [`extra`][pydantic.ConfigDict.extra] configuration value
```

File: docs/concepts/dataclasses.md (hunk 3 above)
Comment (sydney-runkle) [reply docs/concepts/dataclasses.md:43]: This is confusing, but the answer is this happens for both

File: docs/concepts/dataclasses.md (hunk 2 above)
Comment (sydney-runkle) [reply docs/concepts/dataclasses.md:140]: I'm fine with this as long as we link to model docs. Maybe we could add a few other notes as well like json schema, etc.

**Review by sydney-runkle — COMMENTED** [review]
Nice work! Left a few quick notes :)

**Review by sydney-runkle — APPROVED** [review]
Looking great otherwise, happy to merge after the below tweak

**Pull request merged**